packages/core-data/src/dynamic-entities.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * Internal dependencies
+ */
+import type { GetRecordsHttpQuery, State } from './selectors';
+import type * as ET from './entity-types';
+
+export type WPEntityTypes< C extends ET.Context = 'edit' > = {
+	Comment: ET.Comment< C >;
+	Media: ET.Attachment< C >;
+	Menu: ET.NavMenu< C >;
+	MenuItem: ET.NavMenuItem< C >;
+	MenuLocation: ET.MenuLocation< C >;
+	Plugin: ET.Plugin< C >;
+	PostType: ET.Type< C >;
+	Revision: ET.PostRevision< C >;
+	Sidebar: ET.Sidebar< C >;
+	Site: ET.Settings< C >;
+	Status: ET.PostStatusObject< C >;
+	Taxonomy: ET.Taxonomy< C >;
+	Theme: ET.Theme< C >;
+	UnstableBase: ET.UnstableBase< C >;
+	User: ET.User< C >;
+	Widget: ET.Widget< C >;
+	WidgetType: ET.WidgetType< C >;
+};
+
+/**
+ * A simple utility that pluralizes a string.
+ * Converts:
+ * - "post" to "posts"
+ * - "taxonomy" to "taxonomies"
+ * - "media" to "mediaItems"
+ * - "status" to "statuses"
+ */
+type PluralizeEntity< T extends string > = T extends 'Media'
+	? 'MediaItems'
+	: T extends 'Status'
+	? 'Statuses'
+	: T extends `${ infer U }y`
+	? `${ U }ies`
+	: `${ T }s`;
+
+/**
+ * A simple utility that singularizes a string.
+ *
+ * Converts:
+ * - "posts" to "post"
+ * - "taxonomies" to "taxonomy"
+ * - "mediaItems" to "media"
+ * - "statuses" to "status"
+ */
+type SingularizeEntity< T extends string > = T extends 'MediaItems'
+	? 'Media'
+	: T extends 'Statuses'
+	? 'Status'
+	: T extends `${ infer U }ies`
+	? `${ U }y`
+	: T extends `${ infer U }s`
+	? U
+	: T;
+
+export type SingularGetters = {
+	[ Key in `get${ keyof WPEntityTypes }` ]: (
+		state: State,
+		id: number | string,
+		query?: GetRecordsHttpQuery
+	) => WPEntityTypes[ Key extends `get${ infer E }` ? E : never ] | undefined;
+};
+
+export type PluralGetters = {
+	[ Key in `get${ PluralizeEntity< keyof WPEntityTypes > }` ]: (
+		state: State,
+		query?: GetRecordsHttpQuery
+	) => Array<
+		WPEntityTypes[ Key extends `get${ infer E }`
+			? SingularizeEntity< E >
+			: never ]
+	> | null;
+};
+
+type ActionOptions = {
+	throwOnError?: boolean;
+};
+
+type DeleteRecordsHttpQuery = Record< string, any >;
+
+export type SaveActions = {
+	[ Key in `save${ keyof WPEntityTypes }` ]: (
```

Comment:
This still feels a bit limiting since it only targets the entities that core registers. 

What about all the custom post types and taxonomies that are registered and retrieved dynamically? 

See:

https://github.com/WordPress/gutenberg/blob/bea1374a263f7e65702c508c2f715dcd649320f4/packages/core-data/src/entities.js#L220-L230

Reply:
That happens asynchronously and isn't necessarily available when the actions/selectors are registered.

Here is what `wp.data.select('core');` shows in console even though I have WooCommerce active with a CPT named `Product`

<img width="508" alt="image" src="https://github.com/user-attachments/assets/c6db9189-dd60-493f-b1ef-82a9c7b92603" />

Reply:
In fact, those dynamic entities are filtered out here
https://github.com/WordPress/gutenberg/blob/8dd99ede394c871a94bafc33a71abb0cf24c7bc8/packages/core-data/src/index.js#L30

Reply:
Thanks for confirming, that makes sense 👍